### PR TITLE
Initial implementation of ImplicitFuncComp

### LINKED ISSUE #2320: Continue implementation of POEM 057 - ImplicitFuncComp
### Summary of Issue

Add an implementation of ImplicitFuncComp.

### Issue Type

- [ ] Bug
- [x] Enhancement  
      [POEM_057](https://github.com/OpenMDAO/POEMs/blob/master/POEM_057.md)

### PR DESCRIPTION
**Pull request by naylor-b**
### Summary

This continues the implementation of POEM 057 by adding ImplicitFuncComp.  This PR is based off of PR #2309 so that PR will show up in this one until that one is merged.  It may be advisable to wait to review this one until that merge is complete.

### Related Issues

- Resolves #2320

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/components/func_comp_common.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+"""
+Define functions and objects common to the ExplicitFuncComp and ImplicitFuncComp classes.
+"""
+
+try:
```
Comment (Kenneth-T-Moore) [inline openmdao/components/func_comp_common.py:5]: Not sure I understand the jax block. Are users going to import `jax`, `jnp`, etc. from components.fun_comp_common?  Or are they going to import this flag "jax" that tells them whether jax is installed correctly?

File: openmdao/components/implicit_func_comp.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
+"""Define the ImplicitFuncComp class."""
+
+
+import numpy as np
+from numpy import asarray, isscalar
+from itertools import chain
```
Comment (Kenneth-T-Moore) [inline openmdao/components/implicit_func_comp.py:6]: itertools import should be above numpy for better PEP.

File: openmdao/components/implicit_func_comp.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        if solve_nonlinear:
+            self.solve_nonlinear = self._solve_nonlinear_
+        if linearize:
+            self.linearize = self._linearize_
+        if solve_linear:
+            self.solve_linear = self._solve_linear_
+
+    def setup(self):
+        """
+        Define out inputs and outputs.
```
Comment (Kenneth-T-Moore) [inline openmdao/components/implicit_func_comp.py:70]: out -> our

File: openmdao/components/implicit_func_comp.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                kwargs = _copy_with_ignore(meta, omf._allowed_declare_options_args,
+                                           ignore=optignore)
+                self.options.declare(name, **kwargs)
+            else:
+                kwargs = _copy_with_ignore(meta, omf._allowed_add_input_args)
+                self.add_input(name, **kwargs)
+
+        for i, (name, meta) in enumerate(self._apply_nonlinear_func.get_output_meta()):
+            if name is None:
+                raise RuntimeError(f"{self.msginfo}: Can't add output corresponding to return "
```
Comment (Kenneth-T-Moore) [inline openmdao/components/implicit_func_comp.py:86]: Not clear this error has test coverage.

EDIT - found a few more; I guess this initial implementation doesn't yet have the tests for the error messages.

File: openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py [hunk 5]
```diff
@@ -141,11 +148,9 @@ def test_assert(self):
                         if "allow-assert" in tags:
                             continue
 
-                    code = ''.join(block['source'])
-                    if 'assert' in code:
-
-                        msg = f"Assert found in a code block in {file}. "
-                        self.fail(msg)
+                    for line in block['source']:
```
Comment (Kenneth-T-Moore) [inline openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py:151]: How much speedup in the doc build did we get from the changes?

**Review by Kenneth-T-Moore — CHANGES_REQUESTED** [review]

File: openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py (hunk 5 above)
Comment (naylor-b) [reply openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py:151]: My motivation for making changes was to remove the os.walk from the top level scope and put it in a function so it wouldn't execute on import.  Not a huge deal with a test file I suppose, but I fixed it anyway, and while I was in there I saw a few other things that I thought could be improved.  Sometimes I get a little carried away. If you want me to revert the changes I will.  Speed wise I would expect the new version to be slightly faster but I  didn't spend the time to benchmark it.

File: openmdao/components/func_comp_common.py (hunk 1 above)
Comment (naylor-b) [reply openmdao/components/func_comp_common.py:5]: This block is a leftover and isn't needed any more.  I'll remove it.

File: openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py (hunk 5 above)
Comment (Kenneth-T-Moore) [reply openmdao/docs/openmdao_book/tests/test_jupyter_output_linting.py:151]: No, I think this is fine. I just thought you'd found a bottleneck for how slow it builds.

File: openmdao/components/implicit_func_comp.py (hunk 3 above)
Comment (naylor-b) [reply openmdao/components/implicit_func_comp.py]: fixed

File: openmdao/components/implicit_func_comp.py (hunk 2 above)
Comment (naylor-b) [reply openmdao/components/implicit_func_comp.py:3]: fixed

**Review by Kenneth-T-Moore — APPROVED** [review]

File: openmdao/components/tests/test_implicit_func_comp.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            J['R_y2','y2'] = -1
+            J['R_y2','z'] = [1, 1]
+            J['R_y2','y1'] = 0.5*y1**-0.5
+
+        p_opt = om.Problem()
+
+        p_opt.model = om.ImplicitFuncComp(f, linearize=linearize) #, solve_nonlinear=solve_nonlinear)
+
+        newton = p_opt.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False)
+        newton.options['iprint'] = 0
```
Comment (swryan) [inline openmdao/components/tests/test_implicit_func_comp.py:102]: couldn't you do this in one line by passing `iprint` as an arg as well?
`p_opt.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False, iprint=0)`

File: openmdao/components/tests/test_implicit_func_comp.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            J['y2','z'] = [1, 1]
+            J['y2','y1'] = 0.5*y1**-0.5
+
+            J['R_y2','y2'] = -1
+            J['R_y2','z'] = [1, 1]
+            J['R_y2','y1'] = 0.5*y1**-0.5
+
+        p_opt = om.Problem()
+
+        p_opt.model = om.ImplicitFuncComp(f, linearize=linearize) #, solve_nonlinear=solve_nonlinear)
```
Comment (swryan) [inline openmdao/components/tests/test_implicit_func_comp.py:99]: commented code fragment

File: openmdao/components/tests/test_implicit_func_comp.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+                .declare_partials(of='*', wrt='*', method='cs')
+                )
+
+        p = om.Problem()
+        p.model.add_subsystem('comp', om.ImplicitFuncComp(f))
+
+        # need this since comp is implicit and doesn't have a solve_linear
+        p.model.linear_solver = om.DirectSolver()
+        newton = p.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False)
+        newton.options['iprint'] = 0
```
Comment (swryan) [inline openmdao/components/tests/test_implicit_func_comp.py:144]: couldn't you do this in one line by passing `iprint` as an arg as well?
`p.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False, iprint=0)`

File: openmdao/func_api.py [hunk 9]
```diff
@@ -304,7 +316,7 @@ def get_input_meta(self):
         """
         if self._call_setup:
             self._setup()
-        return self._inputs.items()
+        return [it for it in self._inputs.items() if 'resid' not in it[1]]
```
Comment (swryan) [inline openmdao/func_api.py:319]: docstring still says this returns an iterator (now returns a list)

File: openmdao/func_api.py [hunk 10]
```diff
@@ -555,6 +569,26 @@ def _compute_out_shapes(self, ins, outs):
             for name in need_shape:
                 outs[name]['shape'] = shape
 
+    def _default_to_shape(self, name, meta, defaults_dict):
```
Comment (swryan) [inline openmdao/func_api.py:572]: a docstring might be good

Comment (naylor-b) [reply openmdao/func_api.py:572]: fixed

File: openmdao/func_api.py (hunk 9 above)
Comment (naylor-b) [reply openmdao/func_api.py:319]: fixed

File: openmdao/components/tests/test_implicit_func_comp.py (hunk 8 above)
Comment (naylor-b) [reply openmdao/components/tests/test_implicit_func_comp.py]: fixed

File: openmdao/components/tests/test_implicit_func_comp.py (hunk 7 above)
Comment (naylor-b) [reply openmdao/components/tests/test_implicit_func_comp.py]: fixed

File: openmdao/components/tests/test_implicit_func_comp.py (hunk 6 above)
Comment (naylor-b) [reply openmdao/components/tests/test_implicit_func_comp.py]: fixed

File: openmdao/components/tests/test_implicit_func_comp.py [hunk 11]
```diff
@@ -98,7 +137,7 @@ def linearize(z, x, y1, y2, R_y1, R_y2, J):
 
         p_opt.model = om.ImplicitFuncComp(f, linearize=linearize)
 
-        newton = p_opt.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False, iprint=0)
+        p_opt.model.nonlinear_solver = om.NewtonSolver(solve_subsystems=False, iprint=0)
```
Comment (swryan) [inline openmdao/components/tests/test_implicit_func_comp.py:140]: Note:  if you think it worthwhile to change, there is similar code in the doc:
(openmdao/docs/openmdao_book/features/building_blocks/components/implicit_func_comp.ipynb)

Comment (naylor-b) [reply openmdao/components/tests/test_implicit_func_comp.py:140]: fixed

**Review by Kenneth-T-Moore — CHANGES_REQUESTED** [review]
Putting a hold while Bret works on the test coverage.

**Review by Kenneth-T-Moore — APPROVED** [review]

**Review by robfalck — APPROVED** [review]

**Pull request merged**